### PR TITLE
Add support for outputting dependency manifest files, used by ninja and make

### PR DESCRIPTION
**Pull request by rgeary1**
Use --manifest-file=somefile.d to output the dependency manifest.
This file will contain a list of files which were read by protoc as part
of creating the output files.  It doesn't include the plugin inputs if
plugins are used, that could be a later extension.
The manifest file is in the format <output file>: <input files>.  The
manifest file format only allows you to specify one output file, which
isn't a problem as it's used to detect input changes in order to detect
when to rerun the protoc command.  The output file used in the manifest
is the manifest filename itself; to use this in ninja you should declare
the manifest file as the first output as well as the depfile input.

### CODE REVIEW

**Comment by xfxyjwf** [comment]
@TeBoring to take a look. We got internal feature requests for this as well. Need to find a solution that works both for internal users and opensource users.

**Comment by TeBoring** [comment]
I don't quite understand what exactly do you want.
Could you give several examples showing what the output file look like given input?

File: src/google/protobuf/compiler/command_line_interface.cc [hunk 1]
```diff
@@ -1282,6 +1309,62 @@ bool CommandLineInterface::GenerateOutput(
...
+    return false;
+  }
+
+  stringstream ss;
+  string output_filename = manifest_name_;
+  if (output_filename.compare(0, 2, "./") == 0) {
+    output_filename = output_filename.substr(2);
+  }
+  ss << output_filename << ": ";
+  for (set<const FileDescriptor*>::const_iterator it = already_seen.begin(); it != already_seen.end(); ++it ) {
```
Comment (TeBoring) [inline src/google/protobuf/compiler/command_line_interface.cc]: Should it iterate on file_set.file()?

Comment (rgeary1) [inline src/google/protobuf/compiler/command_line_interface.cc]: Yes this should be file_set.  Originally I was calling GetSourceLocation, which is a private member of FileDescriptor.

**Comment by rgeary1** [comment]
The makefile depency file output contents is

    <path to output file>: <path to input file1> <path to input file2> ...

you can separate lines for readability with a \ character followed by a newline character

For examples, you can generate this file for a number of situations with gcc with the flags "-MMD -MF test.d".  
The only restriction I know of (for ninja, and maybe make) is there can only be one output file. But since the purpose of the depfile is to invalidate the state of all the outputs if one input changes, then it seems appropriate to give the path to the depfile itself as the output file, and let the build system enumerate the output files.

**Comment by TeBoring** [comment]
Could you sign the cla first? Thanks

**Comment by TeBoring** [comment]
Internally, our need is a little bit different. We don't give protoc the name for manifest file. protoc will generated a .d file for each input file, which records dependencies of just that single input file.
I have also added your feature into that implementation. So, if you use -m, protoc will generate .d file for each proto file. If you use -mFILE, protoc will write all dependencies into FILE.
I have also mentioned this pull request in my code. You don't need to change your code, but I still need your CLA. Does it sound good to you?

**Comment by rgeary1** [comment]
Sounds good.  Out of interest, which directory do you output the files to
if you just specify -m?  I can't see your pull request - does the depfile
generated by your code look identical to mine?
I'm working on getting the CLA, I developed this as part of my employment
so I presume our company needs to sign it.  Should happen, just will take
some time. My guess is a week.

**Comment by TeBoring** [comment]
If you just have -m, it will generate file in the same directory as the input file.
The content's format should be the same. I implemented internally, but it will be visible in our next opensource release.
But I have a question, do you expect all paths to be absolute paths in the following example?

    <path to output file>: <path to input file1> <path to input file2> ...

**Comment by rgeary1** [comment]
We have many users who have read-only access to some source directories.  We're not intending to use the -m multi-file option, but you may want to consider making the multi-output directory explicit.

Depfile syntax : The output path should match the format expected by the build system, which will only be visible via the command line. If it's specified as relative on the cmdline, the output path needs to be relative. If it's specified as absolute on the cmdline it needs to be absolute in the depfile.  It's the same for the input files.

**Comment by TeBoring** [comment]
Some files are not input files, but they are included by other input files.
How about their path?

**Comment by rgeary1** [comment]
The most general solution would be to provide an option to the user. The option would specify a root path for depfile paths, and all paths under the root would be output as relative paths. If the option was not specified, the paths would be absolute.

I realize I've just proposed adding several new command line options to protoc; some projects resist command line option bloat, some seem to embrace it (eg. "man gcc"!).  If I had to choose, I'd say use absolute paths always since it's unambiguous.

**Comment by TeBoring** [comment]
I prefer to use absolute path always.

**Comment by TeBoring** [comment]
We just found the feature you implemented is enough for internal use. So instead, I want to clean up this pull request. I sent another pull request: https://github.com/google/protobuf/pull/193 to clean the code.
It pulled your change and made change on it directly. After you signed cla, I will merge #193 instead of #178 .
Please take a look at #193. Thanks.

**Comment by rgeary1** [comment]
Signed the CLA, Tower Research Capital

**Comment by TeBoring** [comment]
@rgeary1 Could you add your email (the one associated with this pull request) to tower-research-google-open-source-contributors@googlegroups.com

**Comment by rgeary1** [comment]
Both my work & gmail email address are already members in the google group. The commit email is my work email.  I've tried adding my work email to my github account.  Do you know which email I need to add to the group?

**Pull request merged**